src/Microsoft.AspNetCore.JsonPatch/Helpers/ObjectTreeAnalyisResult.cs

```diff
@@ -127,12 +138,15 @@ internal class ObjectTreeAnalysisResult
             // case, it's valid for add if there's 1 item left in the propertyPathTree.
             //
             // it can also be a property info.  In that case, if there's nothing left in the path
-            // tree we're at the end, if there's one left we can try and set that. 
-            if (targetObject is IDictionary<string, object>)
+            // tree we're at the end, if there's one left we can try and set that.
+            var idictionaryType = ClosedGenericMatcher.ExtractGenericInterface(
```

Comment:
There is duplication of logic in this method which we can cleanup later (I have a separate set of changes which are bigger and probably need some discussion before we agree to proceed with them), but for now I tried to do minimal set of changes to fix this issue.